conjure-runtime/src/client_factory.rs

```diff
@@ -365,4 +341,129 @@ impl ClientFactory {
             handle: self.0.cache_manager.uncached().blocking_handle.clone(),
         })
     }
+
+    /// Creates a refreshable collection of clients, each corresponding to a separate replica of the
+    /// service.
+    ///
+    /// # Note
+    ///
+    /// The client type `T` is assumed to be stateless - each instance will be recreated on every
+    /// refresh.
+    pub fn per_host_clients<T>(
+        &self,
+        service: &str,
+    ) -> Result<Refreshable<PerHostClients<T>, Error>, Error>
+    where
+        T: AsyncService<Client> + 'static + Sync + Send,
+    {
+        self.per_host_clients_inner(service, T::new)
+    }
+
+    /// Creates a refreshable collection of blocking clients, each corresponding to a separate
+    /// replica of the service.
+    ///
+    /// # Note
+    ///
+    /// The client type `T` is assumed to be stateless - each instance will be recreated on every
+    /// refresh.
+    pub fn blocking_per_host_clients<T>(
+        &self,
+        service: &str,
+    ) -> Result<Refreshable<PerHostClients<T>, Error>, Error>
+    where
+        T: Service<blocking::Client> + 'static + Sync + Send,
+    {
+        self.per_host_clients_inner(service, {
+            let handle = self.0.cache_manager.uncached().blocking_handle.clone();
+            move |client| {
+                T::new(blocking::Client {
+                    client,
+                    handle: handle.clone(),
+                })
+            }
+        })
+    }
+
+    fn per_host_clients_inner<T>(
+        &self,
+        service: &str,
+        make_client: impl Fn(Client) -> T + 'static + Sync + Send,
+    ) -> Result<Refreshable<PerHostClients<T>, Error>, Error>
+    where
+        T: 'static + Sync + Send,
+    {
+        let state_builder = self.state_builder(service);
+
+        self.0
+            .config
+            .map({
+                let service = service.to_string();
+                move |c| c.merged_service(&service).unwrap_or_default()
+            })
+            .try_map({
+                move |c| {
+                    let mut clients = HashMap::new();
+
+                    for (i, uri) in c.uris().iter().enumerate() {
+                        let host = Host { uri: uri.clone() };
+                        let host_config = service_config::Builder::from(c.clone())
+                            .uris([uri.clone()])
+                            .build();
+
+                        let state = state_builder.build(&host_config, Some(i))?;
+                        let state = Arc::new(ArcSwap::new(state));
+                        let client = make_client(Client::new(state, None));
```

Comment:
One implication of this setup that I want to call out is that each individual client is not live reloading - changes to e.g. the read timeout will result in new client objects created but the old ones will not update.

We could adjust this logic to make the clients reload internally but I don't know if it's worth the extra complexity.